### PR TITLE
Add support for phantom files (#3956)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Connected issue: #3956 

### Author

- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain                <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                          <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

## Author (hotfixes)

- [x] Added `h` tag to commit title and PR targets `prod`           <sub>or this PR does not include a temporary hotfix</sub>
- [x] Added `H` tag to commit title                                 <sub>or this PR does not include a permanent hotfix</sub>
- [x] Added `hotfix` label to PR                                    <sub>or this PR does not include a hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [ ] ~Ticket is in *Review requested* column~ skipped
- [ ] ~Requested review from peer reviewer~ skipped
- [ ] ~Assigned PR to peer reviewer~ skipped

### Peer reviewer (after approval)

- [x] `N reviews` label on PR reflects prior peer reviews with changes requested
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                         <sub>or PR description does not contain an announcement</sub>
- [ ] Shortened the PR chain                                        <sub>or this PR is not the base of another PR</sub>
- [ ] Pushed merge commit to Gitlab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab
- [ ] Moved connected issue to `prod` or `Merged prod`              <sub>or this PR does not represent a promotion</sub>

### Operator (reindex) 

- [ ] Started reindex in target deployment                          <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>
- [ ] Filed backport PR                                             <sub>or this PR does not represent a hotfix</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: scripts/recan_bundle_tdr.py [hunk 1]
```diff
@@ -245,6 +245,9 @@ def to_json_row(self) -> JSON:
         return dict(super().to_json_row(),
                     file_id=drs_uri(self.file_manifest_entry['drs_path']),
                     descriptor=dict(tdr.Checksums.from_json(self.file_manifest_entry).to_json(),
+                                    # FIXME: Support non-null DRS URIs in file descriptors
+                                    #        https://github.com/DataBiosphere/azul/issues/3631
+                                    drs_uri=None,
```
Comment (hannes-ucsc) [inline scripts/recan_bundle_tdr.py:250]: This looks as if it makes every file in the can is now a phantom file. That
s not representative of reality. Then default case should be normal, internal file.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I don't think this is complete. For example, I would expect the signature to `RepositoryPlugin.drs_uri` to be updated to reflect `drs_path` being None. I am on the fence how much effort we should put in to unit test this special case against the callsites to that plugin method, but we should try.

To test this effectively we need to add a lungmap catalog in `dev` so I will be adding a ticket to have the LungMap snapshots copied to TDR `dev` and another one to add the catalog to Azul `dev`.

**Comment by noah-aviel-dove** [comment]
Tested using the catalog configuration shown [here](https://github.com/DataBiosphere/azul/pull/4034/files)

File: src/azul/plugins/repository/canned/__init__.py [hunk 2]
```diff
@@ -138,7 +138,8 @@ def fetch_bundle(self, bundle_fqid: CannedBundleFQID) -> Bundle:
                  time.time() - now, bundle.uuid, bundle.version)
         return bundle
 
-    def drs_uri(self, drs_path: str) -> str:
+    def drs_uri(self, drs_path: Optional[str]) -> str:
+        assert drs_path is not None
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/canned/__init__.py:142]: Implement phantom files in kind for these other plugins.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 3]
```diff
@@ -741,12 +747,24 @@ def _add_manifest_entry(self,
             )
         })
 
-    def _parse_file_id_column(self, file_id: Optional[str]) -> Optional[str]:
+    def _parse_drs_uri(self,
+                       file_id: Optional[str],
+                       descriptor: JSON
+                       ) -> Optional[str]:
+        try:
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:754]: If TDR somehow provides a `file_id` we should honor and use that. We should only fall back to the external DRS URI if TDR doesn't provide one. This wasn't spec'ed anywhere.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**